File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
int f(vector<vector<int>>& dp,vector<int>& nums,int prev,int i) {
    if(i==nums.size()) return 0;
    if(dp[i][prev+1]!=-1) return dp[i][prev+1];
    int pick=0;
    int npick=0;
    if(prev==-1 || nums[i]>nums[prev]) {
        pick = 1 + f(dp,nums,i,i+1);
    }
    npick = 0 + f(dp,nums,prev,i+1);
    return dp[i][prev+1] = max(pick,npick);
}
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n+1,vector<int>(n+1,-1));
        return f(dp,nums,-1,0);
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1]*(n+1)
        def helperFunction(index):
            #Whats the length of longest increasing subsequence starting at index i?
            if index>=n: return 0
            if dp[index]!=-1: return dp[index]
            maxLengthStartingAtI = 1
            for j in range(index+1,n):
                if nums[j]>nums[index]:
                    maxLengthStartingAtI = max(maxLengthStartingAtI,
                    helperFunction(j)+1)
            dp[index] = maxLengthStartingAtI
            return dp[index]
        
        for i in range(len(nums)): helperFunction(i)
        return max(dp)
        

        